zeus/device/cpu/rapl.py

```diff
@@ -29,6 +32,94 @@
 
 
+class RaplWraparoundTracker:
+    """Monitor the wrapping around of RAPL counters.
+
+    This class acts as a lower level wrapper around a Python process that polls
+    the wrapping of RAPL counters. This is primarily used by
+    [`RAPLCPUs`][zeus.device.cpu.rapl.RAPLCPUs].
+
+    !!! Warning
+        Since the monitor spawns a child process, **it should not be instantiated as a global variable**.
+        Python puts a protection to prevent creating a process in global scope.
+        Refer to the "Safe importing of main module" section in the
+        [Python documentation](https://docs.python.org/3/library/multiprocessing.html#the-spawn-and-forkserver-start-methods)
+        for more details.
+
+    Attributes:
+        rapl_file_path (str): File path of rapl file to track wraparounds for.
+        max_energy_uj (float): Max value of rapl counter for `rapl_file_path` file. Used to
+            determine the sleep period between polls
+    """
+
+    def __init__(
+        self,
+        rapl_file_path: str,
+        max_energy_uj: float,
+    ) -> None:
+        """Initialize the rapl monitor.
+
+        Args:
+            rapl_file_path: File path where the RAPL file is located
+            max_energy_uj: Max energy range uj value
+        """
+        if not os.path.exists(rapl_file_path):
+            raise ValueError(f"{rapl_file_path} is not a valid file path")
+
+        # Set up logging.
+        self.logger = get_logger(type(self).__name__)
+
+        self.logger.info("Monitoring wrap around of %s", rapl_file_path)
+
+        context = mp.get_context("spawn")
+        self.wraparound_counter = context.Value("i", 0)
+        # Spawn the power polling process.
+        atexit.register(self._stop)
+        self.process = context.Process(
+            target=_polling_process,
+            args=(rapl_file_path, max_energy_uj, self.wraparound_counter),
+        )
+        self.process.start()
+
+    def _stop(self) -> None:
+        """Stop monitoring power usage."""
+        if self.process is not None:
+            self.process.terminate()
+            self.process.join(timeout=1.0)
+            self.process.kill()
+            self.process = None
+
+    def get_num_wraparounds(self) -> int:
+        """Get the number of wraparounds detected by the polling process."""
+        with self.wraparound_counter.get_lock():
+            return self.wraparound_counter.value
+
+
+def _polling_process(
+    rapl_file_path: str,
+    max_energy_uj: float,
+    wraparound_counter,
+) -> None:
+    """Check for wraparounds in the specified rapl file."""
+    try:
+        # Use line buffering.
+        with open(rapl_file_path, "r") as rapl_file:
+            last_energy_uj = float(rapl_file.read().strip())
+        while True:
+            sleep_time = 1.0
+            with open(rapl_file_path, "r") as rapl_file:
+                energy_uj = float(rapl_file.read().strip())
+            if max_energy_uj - energy_uj < 1000:
```

Comment:
What was your basis of setting this value to 1000? If someone gets unlucky, can this be a bad value?

Reply:
I did a quick google search and it seemed like a reasonable maximum CPU power draw to account for was around 200 W and I multiplied it by 5 to be on the safe side. I'm not sure what the power draw for a CPU used during training would be so this value definitely could be underestimated. One issue I could see happening is if it takes a long time to acquire the lock around `wraparound_counter`

Reply:
`RAPLFile` is pretty much a singleton would only hold the lock for a very short time (just reading out the counter and that's it). The only other entity that acquires the lock is the polling process. So I think the lock will rarely take very long to acquire, if ever.

Reply:
But BTW aren't things in the unit of microJoules? If you assumed that 1000 W is the maximum power draw, the counter value will advance by at most 1000 * 1e6 * 0.1 microJoules?

Reply:
AHH, yes it is in microjoules so assuming 1000W max and 0.1 second polling the comparison should be 10^8
```
if max_energy_uj - energy_uj < 100000000:
```

Reply:
Right. Please document the meaning of 1e8 in the source code as well. Probably make it a constant like this:

```
# Assuming this and that...
WRAPAROUND_CLOSE = 1000 * 1e6 * 0.1
```